javascript-style-guide.md

```diff
@@ -2,148 +2,232 @@
-Most text editors can be configured to format whitespace. Please do this to
-avoid whitespace-oriented commits and diffs.
+This helps to avoid whitespace-oriented commits and diffs. Installing an
+[editorconfig plugin][] is one easy way of enabling these rules.
+
+Use JSCS to validate the format of your code. Our formatting rules derive from
+[Code Conventions for the JavaScript Programming Language][]. Your editor may
+offer JSCS integration, which you may find useful if you are not familiar with
+the style guide. You can also use our projects' Grunt or Gulp interfaces.
+
+[Code Conventions for the JavaScript Programming Language]: http://javascript.crockford.com/code.html
+[editorconfig plugin]: http://editorconfig.org/#download
 
-Use JSCS to validate the format of your code. Your editor may offer JSCS
-integration, which you may find useful if you are not familiar with the style
-guide. You can also use our projects' Grunt interface.
+## English
 
-# Grammar
+### Spelling
 
-## Use real words for identifiers.
+Use real words for identifiers.
 
-"element" is best written as "element", not "el", "elm", "elem", or "e".
+For instance, "element" is best written as "element", not "el", "elm", "elem",
+or "e".
 
 Ubiquitous acronyms such as "http" and "url" are permissible.
 
-## Capitalize only the first letter of words in identifiers.
+### Capitalization
 
-A phrase such as "File Transfer URI", which contains a normally-capitalized
-acronym, has the identifier `fileTransferUri` (as opposed to something like
-`fileTransferURI`).
+Capitalize only the first letter of words in identifiers.
 
-# Syntax
+For instance, a phrase such as "File Transfer URI", which contains a
+normally-capitalized acronym, has the identifier `fileTransferUri` (as opposed
+to something like `fileTransferURI`).
 
-## Functions
+### Word choice
 
-Use [function expressions][]. Do not use [function declarations][].
+Prefer descriptiveness, then brevity. Strive for clarity; avoid ambiguity.
 
-[function expressions]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/function
-[function declarations]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function
+Functions are verbs, everything else should be nouns.
 
-Do:
+## Syntax
+
+### Variables
+
+If ECMAScript 6 features are available, prefer `const`. Use `let` only when the
+value needs to change.
+
+`const` and `let` variables should be defined separately, and close to their
+site of first use. `var` statements should be placed at the top of a function.
+
+## Quality
+
+Use JSHint to validate the quality of your code and check for errors.
+
+It is highly recommended that you integrate JSHint into your editor, or switch
+to an editor that offers integration. Real-time error reporting can make you
+more productive by pointing out syntactic and logical errors before runtime.
+
+## Style
+
+### Functionalism
+
+Prefer `forEach`, `map`, `filter` and `reduce` over `for` loops.
```

Comment:
`for` is still a decent way of doing something `n` times, and we [have used it](https://github.com/StoryCloud/we-minder-web/blob/fae9755d5c95b5e5f5cef1cd50fa72e7805afafc/application/common/resources/image-processor-service.js#L32) (and should use it) in cases where it could possibly improve performance. (At least until tail recursion becomes widely available.)

It turns out that most of the time `for` is just being used to iterate over something. C has conditioned us to think it's the right way to iterate, but it turns out that iteration can be abstracted further than `for`.

```js
var array = [1, 2, 3];
var i;
for (i = 0; i < array.length; i += 1) {
    element = array[i];
    // ...
}
```

Almost every time you write a `for` loop, you recreate the same structure. Initialize, check bounds, increment, get an element. `forEach` abstracts all of that, so you can focus on more important logic.

```js
var array = [1, 2, 3];
array.forEach(function (element) {
    // ...
});
```